### ISSUE TITLE
KFolditerator makes a copy of the dataset. 

### ISSUE THREAD

**Issue by RobAltena** [deeplearning4j/deeplearning4j#6784]
Our implementation of KFolditerator makes a copy of the dataset:
https://github.com/deeplearning4j/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/iterator/KFoldIterator.java#L54

SKlearn creates an array of indices to shuffle. It looks more elegant.
https://github.com/scikit-learn/scikit-learn/blob/55bf5d9/sklearn/model_selection/_split.py#L356

Would it make sense to re-implement the dl4j kfolditerator in the spirit of sklearn?

Eventually this will result in a kfolditerator sample: 
https://github.com/deeplearning4j/dl4j-examples/issues/633

**Comment by AlexDBlack** [comment]
The biggest problem with the k-fold iterator IMO is that it requires all data in memory, which significantly restricts its usability.
That's not really fixable until I get back to this datavec rewrite though...

Though at alternative approach might be to save the underlying iterator as DataSets on disk and read in only what we need. But that won't work for something like imagenet, as the images in DataSet format (i.e., float32 or even uint8) is way bigger file size than the original (jpg/png etc) images.

**Comment by RobAltena** [comment]
Thanks Alex. I will stick to a simple k-fold iterator example.

**Comment by printomi** [comment]
Hi, I think that the copy operation of the DataSet is not neccessary here:
https://github.com/deeplearning4j/deeplearning4j/blob/41f220fbc2410efe71d535f8344f74e11ea8a44a/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/iterator/KFoldIterator.java#L54
After all, since KFoldIterator is an iterator, I don't expect that it makes a copy of the DataSet it iterates on. 
What do you think?

**Comment by RobAltena** [comment]
I like the way sklearn just juggles with indices. But as @AlexDBlack pointed out: 

> The biggest problem with the k-fold iterator IMO is that it requires all data in memory, which significantly restricts its usability.

So I would be happy with some sensible improvement of the current KFoldIterator, but would hold off a little on a major re-write. This is why I ended up writing a really simple example.

**Comment by printomi** [comment]
I understand that it would be good to have a KFold solution that uses indices and does not copy data at all.
However, what I was asking about is an easy-to-do improvement that I could do right now, in #6810 .
That line https://github.com/deeplearning4j/deeplearning4j/blob/41f220fbc2410efe71d535f8344f74e11ea8a44a/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/iterator/KFoldIterator.java#L54
could be replaced by 
```java
this.allData = allData;
```
We don't lose anything, but we could save a full copy of the DataSet in memory.
What do you think? Are there any disadvantages of this change?

**Comment by printomi** [comment]
OK, I have eliminated the mentioned copy. One copy remains though: the `train` and `test` fields of `KFoldIterator`.
See commit https://github.com/deeplearning4j/deeplearning4j/pull/6810/commits/6420349c7c42defa74fe4263c621df0cb1bd1e3d
